### ISSUE TITLE
A question about the weighted-sum in GDAS

### ISSUE THREAD

**Issue by liuzili97** [D-X-Y/NAS-Projects#36]
Hi, thanks for your great work. But I have a question about the weighted-sum in GDAS.
In `lib/models/cell_searchs/search_cells.py, line58`, the weighted-sum is calculated as follows:
`weigsum  = sum( weights[_ie] * edge(nodes[j]) if _ie == argmaxs else weights[_ie] for _ie, edge in enumerate(self.edges[node_str]) )`
I don't get it why we need to add the `weights[_ie]` when `_ie != argmaxs`. I think the `weigsum = weights[_ie] * edge(nodes[j])` where `_ie == argmaxs` can be fine. Is there any difference that I don't notice?

**Comment by D-X-Y** [comment]
In the forward, they are the same. However, in the backward, these two kinds of codes are different.
`weights[_ie] * edge(nodes[j]) if _ie == argmaxs else weights[_ie]` allows the gradient BP to all architecture parameters instead of only the argmax one.

**Comment by liuzili97** [comment]
Thanks for your reply!

But even though the `weights[_ie]` is not added when `_ie != argmaxs`, I think the gradient can still pass though all the architecture parameters. This is because `hardwts = one_h - probs.detach() + probs` and `probs   = nn.functional.softmax(logits, dim=1)`.

Have you ever experimented without adding the `weights[_ie]` when `_ie != argmaxs`, and is there a difference in accuracy?

**Comment by liuzili97** [comment]
Thanks for your reply!

But even though the weights[_ie] is not added when _ie != argmaxs, I think the gradient can still pass though all the architecture parameters. This is because hardwts = one_h - probs.detach() + probs and probs = nn.functional.softmax(logits, dim=1).

**Comment by D-X-Y** [comment]
Yes, it does. But it is different from "BP to all weights then to all architecture-parameters".
With this style `weights[_ie] * edge(nodes[j]) if _ie == argmaxs else weights[_ie]`, the gradients are more "correct".

**Comment by liuzili97** [comment]
Ok, thanks!

**Comment by aragakiyuiii** [comment]
Hello, thanks for your great work! I have run your code and have no bug and trouble. 
But I want to ask you a question: In `lib/models/cell_searchs/search_cells.py, line58`, the weighted-sum in Gumbel-softmax is calculated as follows: `weigsum = sum( weights[_ie] * edge(nodes[j]) if _ie == argmaxs else weights[_ie] for _ie, edge in enumerate(self.edges[node_str]) )`. `weights[_ie] * edge(nodes[j])` is tensor but `weights[_ie]` is scalar. `weights[_ie]` does not multiply with a tensor like `edge(nodes[j])`. So `weights[_ie] * edge(nodes[j])` and `weights[_ie]` do not have the same dimension. 

However, code`weigsum = sum( weights[_ie] * edge(nodes[j]) if _ie == argmaxs else weights[_ie] for _ie, edge in enumerate(self.edges[node_str]) )` is all right when running. Why `weights[_ie] * edge(nodes[j])` and `weights[_ie]` can be added together even if their dimensions(data.shape) are different ?
